### ISSUE TITLE
data loader cache shared between mutation loading and mutation resolving

### ISSUE THREAD

**Issue by stevecrozz** [rmosolgo/graphql-ruby#4809]
**Describe the bug**

I raised this issue in #4600 and it has been partially addressed. Previously I had to run `context.dataloader.instance_variable_get(:@source_cache).each_value(&:clear)` in order to clear the dataloader cache, but now I have a nicer interface with `context.dataloader.clear_cache`.

But it is still the case that I have to run this in my mutation resolvers because any object loaded via dataloader, either for loading arguments or using dataloaders in authorization, remains cached all the way through the ultimate field resolution. This is a problem for mutations that expose a field for the same object that was loaded because I cannot access the post-mutation state for any such objects.

For example, I may have a `SetPostName` mutation that loads a post object and then exposes the updated post as a field. If I use a dataloader to load that post, then I have to clear the cache before resolving any of the fields or I'll get the original post object with the old name.

So basically, what I end up with is a mutation with a resolve method like this:
~~~ruby
def resolve
  context.dataloader.clear_cache
  {
    object: context.dataloader.with(...).load(...)
  }
end
~~~

**Versions**

`graphql` version: 2.2.5

**Comment by rmosolgo** [comment]
Great point -- thanks for the detailed writeup!

I think my previous change would have "fixed" this for any fields coming _after_ `def resolve` -- but the problem holds when the object is loaded as an argument, then modified in the resolve body, then re-loaded (but not actually, because of the cache) via dataloader to build a response. 

I agree this is basically a bug, but out of curiosity, what about passing the modified `post` to the result hash instead of re-loading it with Dataloader? (I can imagine several reasons, but if you don't mind sharing some details about your setup, I'm sure I'd be smarter for it!)

**Comment by stevecrozz** [comment]
Returning the modified post works for the first level of the graph. But trouble comes when querying more deeply than the first level. A query within the mutation response for a sub-field like (ie post.author) follows connections between types just as it would in a regular non-mutation query. And within those types, I am extensively using dataloaders to prevent N+1 queries.

If I had already used a dataloader to load that author prior to resolving, perhaps in order to authorize the mutation, then the author I get back in the response will be a pre-mutation view of that author.

At least, this is my understanding of what is happening.

**Comment by rmosolgo** [comment]
I'd _expect_ it to work for those sub-fields, thanks to: 

https://github.com/rmosolgo/graphql-ruby/blob/58b5a3c0ee7b45737e1dd83bb4ff985bf9299f7e/lib/graphql/execution/interpreter/runtime.rb#L225-L229

I wrote up a little replication script which demonstrates the issue you described above (if you uncomment the last bit of the mutation field) but works fine in later fields: 

<details>
<summary>Mutation/Dataloader cache bug</summary>

<p>

```ruby
require "bundler/inline"

gemfile do
  gem "graphql", "2.2.5"
end

class MySchema < GraphQL::Schema
  DATA = {
    "1" => { id: "1", name: "Spatula", other_thing_id: "2" }.freeze,
    "2" => { id: "2", name: "Whisk", other_thing_id: "1"}.freeze,
  }

  class ThingSource < GraphQL::Dataloader::Source
    def fetch(ids)
      ids.map { |id| DATA.fetch(id) }
    end
  end

  class Thing < GraphQL::Schema::Object
    field :name, String
    field :other_thing, Thing

    def other_thing
      dataloader.with(ThingSource).load(object[:other_thing_id])
    end
  end


  class Mutation < GraphQL::Schema::Object
    field :rename_thing, Thing do
      argument :new_name, String
      argument :thing_id, ID, loads: Thing
    end

    def rename_thing(thing:, new_name:)
      thing = thing.dup
      thing[:name] = new_name
      # update the "database"
      DATA[thing[:id]] = thing.freeze
      # Returning this breaks it:
      # dataloader.with(ThingSource).load(thing[:id])
    end
  end

  mutation(Mutation)
  use GraphQL::Dataloader

  def self.object_from_id(id, ctx)
    ctx.dataloader.with(ThingSource).load(id)
  end

  def self.resolve_type(_type, object, ctx)
    Thing
  end
end


pp MySchema.execute(<<-GRAPHQL).to_h
mutation {
  newSpatula: renameThing(thingId: "1", newName: "Pannenlikker") {
    name
    otherThing {
      name
    }
  }
  newWhisk: renameThing(thingId: "2", newName: "Garde") {
    name
    otherThing {
      name
    }
  }
}
GRAPHQL

# {"data"=>
#   # The first bit of the response has the updated name for one Thing 
#   {"newSpatula"=>{"name"=>"Pannenlikker", "otherThing"=>{"name"=>"Whisk"}},
#   # The second bit of the response has updated names for both things:
#    "newWhisk"=>{"name"=>"Garde", "otherThing"=>{"name"=>"Pannenlikker"}}}}
```

</p>
</details>

But if you can share an example where those sub-fields are showing "stale" data from the dataloader cache, I'd love to see it -- there's a lot going on in there and I might have missed a spot!

**Comment by stevecrozz** [comment]
I can try to put together a minimal example later, but let's see if this further info is enough.

I have a mutation with an argument like this:
`argument :my_object_id, type: ID, loads: Types::MyObject`

Types::MyObject defines `self.authorized` which uses context.dataloader to load another object needed to authorize this object. And this other object is still in the dataloader cache when we get to the mutation's resolve method.

**Comment by rmosolgo** [comment]
Oh, I see... yes, I'd certainly believe that. The runtime clears the dataloader cache before and after resolving each mutation, but `loads:` and `def self.authorized?` are part of resolution, from the runtime's perspective. Maybe we could move the before-resolve `clear_cache` _inside_ `GraphQL::Schema::Mutation`, so that it could clear the cache _after_ those helpers are used but _before_ `def resolve` is called. 

That would address mutations inheriting from `GraphQL::Schema::Mutation`, but it wouldn't help mutation fields defined with plain `field  ...`-style definitions, using methods as resolvers. So maybe an improvement to the docs would cover that case for now.
